src/nvim/buffer.c

```diff
@@ -4026,65 +4025,57 @@ char *buf_spname(buf_T *buf)
+    // Reset 'statuscolumn' if it has no dedicated signcolumn but it would have changed in size
+    if (*wp->w_p_stc != NUL && wp->w_minscwidth <= SCL_NO) {
+      wp->w_nrwidth_line_count = 0;
+    }
+    changed_line_abv_curs_win(wp);
+    redraw_later(wp, UPD_NOT_VALID);
   }
 }
 
-/// Invalidate the signcolumn if needed after adding a sign ranging from line1 to line2.
-void buf_signcols_add_check(buf_T *buf, linenr_T line1, linenr_T line2)
+/// "b_signcols" size changed, only redraw entire window if necessary.
+void buf_redraw_signcols(buf_T *buf, int size)
 {
-  if (!buf->b_signcols.sentinel) {
-    return;
-  }
-
-  linenr_T sent = buf->b_signcols.sentinel;
-  if (sent >= line1 && sent <= line2) {
-    // If added sign overlaps sentinel line, increment without invalidating.
-    if (buf->b_signcols.size == buf->b_signcols.max) {
-      buf->b_signcols.max++;
+  buf->b_signcols.size = size;
+  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    if (wp->w_buffer == buf) {
+      win_redraw_signcols(wp);
     }
-    buf->b_signcols.size++;
-    return;
-  }
-
-  if (line1 < buf->b_signcols.invalid_top) {
-    buf->b_signcols.invalid_top = line1;
-  }
-  if (line2 > buf->b_signcols.invalid_bot) {
-    buf->b_signcols.invalid_bot = line2;
   }
 }
 
-int buf_signcols(buf_T *buf, int max)
+/// Update "b_signcols" lines after deleting a sign ranging from "line1" to
+/// "line2" and check if its size changes.
+void buf_signcols_del_check(buf_T *buf, linenr_T line1, linenr_T line2)
 {
-  if (!buf->b_signs_with_text) {
-    buf->b_signcols.size = 0;
-  } else if (max <= 1 && buf->b_signs_with_text >= (size_t)max) {
-    buf->b_signcols.size = max;
-  } else {
-    linenr_T sent = buf->b_signcols.sentinel;
-    if (!sent || max > buf->b_signcols.max) {
-      // Recheck if the window scoped maximum 'signcolumn' is greater than the
-      // previous maximum or if there is no sentinel line yet.
-      buf->b_signcols.invalid_top = sent ? sent : 1;
-      buf->b_signcols.invalid_bot = sent ? sent : buf->b_ml.ml_line_count;
+  for (int i = line1; i <= line2; i++) {
+    int *size = map_ref(int, int)(&buf->b_signcols.lines, i, NULL);
+    if ((*size)-- == buf->b_signcols.size) {
+      buf->b_signcols.valid = false;  // sign deleted from a sentinel line
     }
-
-    if (buf->b_signcols.invalid_bot) {
-      decor_validate_signcols(buf, max);
+    if (*size == 0) {
+      map_del(int, int)(&buf->b_signcols.lines, i, NULL);
     }
   }
+}
 
-  buf->b_signcols.max = max;
-  buf->b_signcols.invalid_top = MAXLNUM;
-  buf->b_signcols.invalid_bot = 0;
-  return buf->b_signcols.size;
+/// Update "b_signcols" lines after adding a sign ranging from "line1" to
+/// "line2" and check if its size changes.
+void buf_signcols_add_check(buf_T *buf, linenr_T line1, linenr_T line2)
+{
+  for (int i = line1; i <= line2; i++) {
+    int *size = map_put_ref(int, int)(&buf->b_signcols.lines, i, NULL, NULL);
```

Comment:
Don't you also have to shift all the lines below downwards? Likewise for deletes you need to shift all the lines below upwards.

Reply:
This is called when signs are added/removed, not lines. But yes the map needs to shifted for added/removed lines, this is done in `extmark_splice_impl()`.